### PR TITLE
Fix build check build submission errors

### PR DESCRIPTION
**Pull request by JaynieBai**
Fixes [#10071](https://github.com/dotnet/msbuild/issues/10071)

### Context
While building with buildcheck enabled and MSBUILDNOINPROCNODE=1
Severity of the rule set to Error, does not reported to the all build:
image 
This is since buildcheck error is not added into the _warningsAsErrorsByProject, WarningsAsErrors == null && _warningsAsErrorsByProject == null is true all the time.   so HasBuildSubmissionLoggedErrors always return false.
https://github.com/dotnet/msbuild/blob/843bfa584a08401b30e4c0b21a3c0b534568877e/src/Build/BackEnd/Components/Logging/LoggingService.cs#L577-L587

Treat warning as errors or message, the buildcheckResultWarning doesn't initialize the code. So when the code of  BuildWarningEventArgs is null. ShouldTreatWarningAsError returns false all the time. 
https://github.com/dotnet/msbuild/blob/a9c95c7251503f85493d3f3bb43780ffa8104589/src/Build/BackEnd/Components/Logging/LoggingService.cs#L1897-L1908
### Changes Made
1. when buildEventArgs is BuildErrorEventArgs, treat  BuildErrorEventArgs' related warnings as errors

2. Initialize the code of BuildCheckResultWarning that is inherited from BuildWarningEventArgs

### Testing
Manually testing on local now

set MSBUILDNOINPROCNODE=1 and change the build_check.BC0101.Severity= Error
dotnet  D:\WORK\msbuild\artifacts\bin\bootstrap\net8.0\MSBuild\MSBuild.dll FooBar.csproj /m:1 -nr:False -restore -analyze
image

Change build_check.BC0101.Severity= warning
dotnet  D:\WORK\msbuild\artifacts\bin\bootstrap\net8.0\MSBuild\MSBuild.dll FooBar.csproj /m:1 -nr:False -restore -analyze -warnaserror
image

### CODE REVIEW

**Comment by JaynieBai** [comment]
Will add a test after figured out the flaky reason of test SampleAnalyzerIntegrationTest https://github.com/dotnet/msbuild/issues/10036

**Review by YuliiaKovalova — APPROVED** [review]

File: src/Build/BackEnd/Components/Logging/LoggingService.cs [hunk 1]
```diff
@@ -1592,6 +1593,12 @@ private void RouteBuildEvent(object loggingEvent)
                 _buildSubmissionIdsThatHaveLoggedErrors.Add(errorEvent.BuildEventContext?.SubmissionId ?? BuildEventContext.InvalidSubmissionId);
             }
 
+            if (buildEventArgs is BuildCheckResultError checkResultError)
```
Comment (JanKrivanek) [inline src/Build/BackEnd/Components/Logging/LoggingService.cs:1596]: We should probably handle the warning + ShouldTreatWarningAsError as well

Comment (JaynieBai) [reply src/Build/BackEnd/Components/Logging/LoggingService.cs:1596]: So there is one condition build with -err:BC0101

Comment (JaynieBai) [reply src/Build/BackEnd/Components/Logging/LoggingService.cs:1596]: Updated with new fix

Comment (JaynieBai) [reply src/Build/BackEnd/Components/Logging/LoggingService.cs:1596]: @JanKrivanek please help review

File: src/Framework/BuildCheck/BuildCheckEventArgs.cs [hunk 2]
```diff
@@ -107,7 +107,8 @@ internal override void CreateFromStream(BinaryReader reader, int version)
 }
 internal sealed class BuildCheckResultWarning : BuildWarningEventArgs
 {
-    public BuildCheckResultWarning(IBuildCheckResult result)
+    public BuildCheckResultWarning(IBuildCheckResult result, string code)
+        : base(subcategory: null, code: code, file: null, lineNumber: 0, columnNumber: 0, endLineNumber: 0, endColumnNumber: 0, message: result.FormatMessage(), helpKeyword: null, senderName: null)
```
Comment (JanKrivanek) [inline src/Framework/BuildCheck/BuildCheckEventArgs.cs:111]: In the future we should populate the location here from the result.Location (that currently is not structured, so bit of refactor will be needed).

Not needed in a scope of this PR

**Review by JanKrivanek — APPROVED** [review]
Looks good! Thank you!

**Pull request merged**